### PR TITLE
Sync unrebased codes

### PR DESCRIPTION
**Pull request by TommyLike**
Changes are below:
1. Add help message for command line.
2. Add enqueue state.
3. Add Retry support
4. Enhance PodGroup support.

**NOTE**: It's depends on the feature in kube-batch, should not be merged before that

The synced codes has not been tested, the e2e testcases would fail due to this.

For #112

### CODE REVIEW

File: pkg/admission/admission_controller.go [hunk 1]
```diff
@@ -62,49 +83,27 @@ func ToAdmissionResponse(err error) *v1beta1.AdmissionResponse {
...
-	var err error
-	policyEvents := map[v1alpha1.Event]struct{}{}
-	exitCodes := map[int32]struct{}{}
+func CheckPolicyDuplicate(policies []v1alpha1.LifecyclePolicy) (string, bool) {
+	policyEvents := map[v1alpha1.Event]v1alpha1.Event{}
+	hasDuplicate := false
+	var duplicateInfo string
 
 	for _, policy := range policies {
-		if policy.Event != "" && policy.ExitCode != nil {
```
Comment (hzxuzhonghu) [inline pkg/admission/admission_controller.go]: wait, why remove the ExitCode?

File: pkg/apis/batch/v1alpha1/job.go [hunk 2]
```diff
@@ -61,13 +61,18 @@ type JobSpec struct {
...
-	//Specifies the queue that will be used in the scheduler, "default" queue is used this leaves empty.
-	Queue string `json:"queue,omitempty" protobuf:"bytes,7,opt,name=queue"`
-
 	// Specifies the plugin of job
 	// Key is plugin name, value is the arguments of the plugin
 	// +optional
 	Plugins map[string][]string `json:"plugins,omitempty" protobuf:"bytes,7,opt,name=plugins"`
+
+	//Specifies the queue that will be used in the scheduler, "default" queue is used this leaves empty.
+	Queue string `json:"queue,omitempty" protobuf:"bytes,7,opt,name=queue"`
```
Comment (hzxuzhonghu) [inline pkg/apis/batch/v1alpha1/job.go]: /hold

the protobuf label 7 is used by plugins already

File: pkg/apis/helpers/helpers.go [hunk 3]
```diff
@@ -65,27 +65,36 @@ func ControlledBy(obj interface{}, gvk schema.GroupVersionKind) bool {
...
+				job.Namespace, job.Name, err)
+			return err
+		}
+		return nil
 	}
 
-	if _, err := kubeClients.CoreV1().ConfigMaps(job.Namespace).Create(cm); err != nil {
-		glog.V(3).Infof("Failed to create ConfigMap for Job <%s/%s>: %v",
+	cmOld.Data = data
+	if _, err := kubeClients.CoreV1().ConfigMaps(job.Namespace).Update(cmOld); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/apis/helpers/helpers.go]: Is there a case we need to update configmap?

File: pkg/apis/helpers/helpers.go [hunk 4]
```diff
@@ -94,6 +103,16 @@ func CreateConfigMapIfNotExist(job *vkv1.Job, kubeClients *kubernetes.Clientset,
 }
 
 func DeleteConfigmap(job *vkv1.Job, kubeClients *kubernetes.Clientset, cmName string) error {
+	if _, err := kubeClients.CoreV1().ConfigMaps(job.Namespace).Get(cmName, metav1.GetOptions{}); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/apis/helpers/helpers.go]: I donot think we need this get ops, it add one more api all

File: pkg/controllers/job/job_controller.go [hunk 5]
```diff
@@ -128,24 +140,23 @@ func NewJobController(config *rest.Config) *Controller {
 	cc.cmdLister = cc.cmdInformer.Lister()
 	cc.cmdSynced = cc.cmdInformer.Informer().HasSynced
 
-	cc.sharedInformers = informers.NewSharedInformerFactory(cc.kubeClients, 0)
```
Comment (hzxuzhonghu) [inline pkg/controllers/job/job_controller.go]: definitely this code is more clean

File: pkg/controllers/job/job_controller.go [hunk 6]
```diff
@@ -154,22 +165,37 @@ func NewJobController(config *rest.Config) *Controller {
...
 // Run start JobController
 func (cc *Controller) Run(stopCh <-chan struct{}) {
 	go cc.jobInformer.Informer().Run(stopCh)
+	go cc.podInformer.Informer().Run(stopCh)
+	go cc.pvcInformer.Informer().Run(stopCh)
 	go cc.pgInformer.Informer().Run(stopCh)
+	go cc.svcInformer.Informer().Run(stopCh)
 	go cc.cmdInformer.Informer().Run(stopCh)
-	go cc.sharedInformers.Start(stopCh)
+	go cc.pcInformer.Informer().Run(stopCh)
```
Comment (hzxuzhonghu) [inline pkg/controllers/job/job_controller.go]: The old ` cc.sharedInformers.Start(stopCh)` is same as these lines.

**Comment by k82cn** [comment]
let's only do code sync up in this PR; we can open other PRs to address comments.

**Comment by TommyLike** [comment]
@k82cn  There are some fundamental issues in master branch of CodeClud repo which lead the half failing of our testcases, for example, the pending state does not exist anymore, needs a fix before sync.

**Comment by k82cn** [comment]
let's make sure code sync up firstly. We can fix the issues in different PRs.

**Pull request closed without merging**